src/libraries/System.Private.CoreLib/src/System/Resources/RuntimeResourceSet.cs

```diff
@@ -309,158 +269,104 @@ private IDictionaryEnumerator GetEnumeratorHelper()
         {
             if (key == null)
                 throw new ArgumentNullException(nameof(key));
-            if (Reader == null || _resCache == null)
+
+            ResourceReader? reader = _defaultReader;
+            Dictionary<string, ResourceLocator>? cache = _resCache;
+            Dictionary<string, ResourceLocator>? caseInsensitiveTable = _caseInsensitiveTable;
+            if (reader is null || cache is null)
                 throw new ObjectDisposedException(null, SR.ObjectDisposed_ResourceSet);
 
-            object? value = null;
-            ResourceLocator resLocation;
+            object? value;
+            ResourceLocator resEntry;
 
-            lock (Reader)
+            lock (cache)
             {
-                if (Reader == null)
-                    throw new ObjectDisposedException(null, SR.ObjectDisposed_ResourceSet);
-
-                if (_defaultReader != null)
+                // Find the offset within the data section
+                int dataPos;
+                if (cache.TryGetValue(key, out resEntry))
                 {
-                    // Find the offset within the data section
-                    int dataPos = -1;
-                    if (_resCache.TryGetValue(key, out resLocation))
-                    {
-                        value = resLocation.Value;
-                        dataPos = resLocation.DataPosition;
-                    }
+                    value = resEntry.Value;
+                    if (value != null)
+                        return value;
 
-                    if (dataPos == -1 && value == null)
-                    {
-                        dataPos = _defaultReader.FindPosForResource(key);
-                    }
+                    // When data type cannot be cached
```

Comment:
Is this true? I am seeing the old code was still doing the following in case dataPos != -1

                        resLocation = new ResourceLocator(dataPos, (ResourceLocator.CanCache(typeCode)) ? value : null);
                        lock (_resCache)
                        {
                            _resCache[key] = resLocation;
                        }

Reply:
Which only re-insert the same resLocation with no reason because value is null

Reply:
ok, but the comment would be confusing as it is already cached with null value. 

Also, I assume dataPos will not equal -1 at that time. is that right? do we need to check for that? or assert at least?

---
In reply to: [520911963](https://github.com/dotnet/runtime/pull/44454#discussion_r520911963) [](ancestors = 520911963)

Reply:
Yes, the old code was reinserting values which couldn't be cached into cache over again. What would you assert for, if `FindPosForResource` returns the negative value the is code which handles that already and the value is never inserted into the cache.

Reply:
The assert I meant is `dataPos != -1`. I guess we shouldn't have as -1 in here. right?

Reply:
No, -1 won't ever be inserted into the cache. See https://github.com/marek-safar/runtime/blob/res/src/libraries/System.Private.CoreLib/src/System/Resources/RuntimeResourceSet.cs#L297